__tests__/atomEffect.test.ts

```diff
@@ -177,6 +177,241 @@ it('should not cause infinite loops when effect updates the watched atom asynchr
+  store.set(watchedAtom, increment)
+  await delay(0)
+  expect(store.get(lettersAndNumbersAtom)).toEqual(['a0', 'b1'])
+  expect(runCount).toBe(4)
+})
+
+it('should allow asynchronous infinite loops with task delay', async () => {
+  expect.assertions(2)
+  let runCount = 0
+  const watchedAtom = atom(0)
+  let done = false
+  const effectAtom = atomEffect((get, { recurse }) => {
+    const value = get(watchedAtom)
+    runCount++
+    if (value >= 3) {
+      done = true
+      return
+    }
+    delay(0).then(() => {
+      recurse(watchedAtom, increment)
+    })
+  })
+  const store = getDefaultStore()
+  store.sub(effectAtom, () => void 0)
+  await waitFor(() => assert(done))
+  expect(store.get(watchedAtom)).toBe(3)
+  expect(runCount).toBe(4)
+})
+
+it('should allow asynchronous infinite loops with microtask delay', async () => {
+  expect.assertions(2)
+  let runCount = 0
+  const watchedAtom = atom(0)
+  watchedAtom.debugLabel = 'watchedAtom' // remove
+  let done = false
+  const effectAtom = atomEffect((get, { recurse }) => {
+    const value = get(watchedAtom)
+    runCount++
+    if (value >= 3) {
+      done = true
+      return
+    }
+    Promise.resolve().then(() => {
+      recurse(watchedAtom, increment)
+    })
+  })
+  const store = getDefaultStore()
+  store.sub(effectAtom, () => void 0)
+  // await waitFor(() => assert(done))
+  done
+  await delay(500)
+  expect(store.get(watchedAtom)).toBe(3)
+  expect(runCount).toBe(4)
+})
+
+it('should work with both recurse and set', async () => {
+  expect.assertions(3)
+  let runCount = 0
+  const watchedAtom = atom(0)
+  const countAtom = atom(0)
+  const effectAtom = atomEffect((get, set) => {
+    const value = get(watchedAtom)
+    get(countAtom)
+    runCount++
+    if (value === 0 || value % 3) {
+      set.recurse(watchedAtom, increment)
+      set(countAtom, increment)
+      return
+    }
+    set(watchedAtom, increment)
+  })
+  const store = getDefaultStore()
+  store.sub(effectAtom, () => void 0)
+  await waitFor(() => assert(store.get(countAtom) === 3))
+  expect(store.get(countAtom)).toBe(3)
+  expect(store.get(watchedAtom)).toBe(4)
+  expect(runCount).toBe(4)
+})
+
+it('should disallow synchronous infinite loops in cleanup', async () => {
+  expect.assertions(3)
+  const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {})
+  let runCount = 0
+  const watchedAtom = atom(0)
+  const anotherAtom = atom(0)
+  const effectAtom = atomEffect((get, { recurse }) => {
+    get(watchedAtom)
+    get(anotherAtom)
+    runCount++
+    return () => {
+      recurse(watchedAtom, increment)
+    }
+  })
+  const store = getDefaultStore()
+  store.sub(effectAtom, () => void 0)
+  await delay(0)
+  store.set(anotherAtom, increment)
+  await delay(0)
+  expect(warnSpy).toHaveBeenCalled()
+  expect(store.get(watchedAtom)).toBe(0)
+  expect(runCount).toBe(2)
+  warnSpy.mockRestore()
+})
+
+// FIXME: is there a way to disallow asynchronous infinite loops in cleanup?
```

Comment:
@dai-shi I couldn't find a way to detect and prohibit this case
```
atomEffect((get, set) => {
  get(watchedAtom)
  return () => {
    setTimeout(() => {
      set.recurse(watchedAtom, increment)
    }, 1000)
  }
})
```

IMO, recursion in the cleanup function is counter-intuitive and should not be supported. The problem I ran into is how, technically, to distinguish it from:
```
atomEffect((get, set) => {
  get(watchedAtom)
  setTimeout(() => {
    set.recurse(watchedAtom, increment)
  }, 1000)
})
```

I was wondering if you had any ideas.